File: Mergesort/mergesort.java
import java.util.*;

class mergesort {
	public static void main(String args[]) {
		ArrayList<Integer> myArr = new ArrayList<Integer>();
		myArr.add(5);
		myArr.add(4);
		myArr.add(3);
		mergeSort(myArr, 0, myArr.size() - 1);
		print(myArr);
	}

	private static <T extends Comparable> void mergeSort(List<T> list, int start, int end) {
		if (start >= end) return;
		int middle = (start + end) / 2;
		mergeSort(list, start, middle);
		mergeSort(list, middle + 1, end);
		merge(list, start, middle, middle + 1, end);

	}

	private static <T extends Comparable> void merge(List<T> list, int leftStart, int leftEnd, int rightStart, int rightEnd) {
		int length = rightEnd - leftStart + 1;
		int left = leftStart;
		int right = rightStart;
		ArrayList<T> temp = new ArrayList<T>();

		for (int i = 0; i < length; i++) {
			if (right > rightEnd) {
				temp.add(list.get(left++));
			}
			else if (left > leftEnd) {
				temp.add(list.get(right++));
			}
			else if (list.get(left).compareTo(list.get(right)) <= 0) {
				temp.add(list.get(left++));
			}
			else {
				temp.add(list.get(right++));
			}
		}

		for (int i = 0; i < length; i++) {
			list.set(leftStart++, temp.get(i));
		}
	}

	private static void print(List<?> list) {
		for (int i = 0; i < list.size(); i++) {
			System.out.print(list.get(i) + " ");
		}
		System.out.println();
	}

}

File: Mergesort/mergesort.go
package main
import "fmt"

func merge (list []int, leftStart int, leftEnd int, rightStart int, rightEnd int) {
	length := rightEnd - leftStart + 1
	temp := make([]int, length)
	left := leftStart
	right := rightStart

	for i := 0; i < length; i++ {
		if (left > leftEnd) {
			temp[i] = list[right]
			right++
		} else if (right > rightEnd) {
			temp[i] = list[left]
			left++
		} else if (list[left] <= list[right]) {
			temp[i] = list[left]
			left++
		} else {
			temp[i] = list[right]
			right++
		}
	}

	for i := 0; i < length; i++ {
		list[leftStart] = temp[i]
		leftStart++
	}
}

func mergesort(list []int, start int, end int) {
	if (end - start < 1) {
		return
	}
	middle := (start + end) / 2

	mergesort(list, start, middle)
	mergesort(list, middle + 1, end)
	merge(list, start, middle, middle + 1, end)
}

func print(list []int) {
	for i := 0; i < len(list); i++ {
		fmt.Print(list[i], " ")
	}
	fmt.Println()
}

func main() {
	list := []int{3, 4, 5, 1, 2, 45, 3, 56}
	mergesort(list, 0, len(list) - 1)
	print(list)
}

